mypy/private/mypy.bzl

```diff
@@ -76,75 +113,83 @@ def _mypy_impl(target, ctx):
     if not hasattr(ctx.rule.files, "srcs"):
         return []
 
-    # we need to help mypy map the location of external deps by setting
-    # MYPYPATH to include the site-packages directories.
-    external_deps = {}
-
     # we need to help mypy map the location of first party deps with custom
     # 'imports' by setting MYPYPATH.
     imports_dirs = {}
 
-    # generated dirs
-    generated_dirs = {}
-
     upstream_caches = []
 
     types = []
 
     depsets = []
 
+    # Since we can't pass a {label: label} mapping as Bazel attr to the aspect
+    # config, it gets passed as a list of keys and a list of vals which we have
+    # to rezip into a mapping. Which is this.
+    #
+    # Reconstructs a mapping of requirements to at most one type provider for
+    # that requirement.
     type_mapping = dict(zip([k.label for k in ctx.attr._types_keys], ctx.attr._types_values))
-    dep_with_stubs = [_.label.workspace_root + "/site-packages" for _ in ctx.attr._types_keys]
+
+    # Look up type providers for the deps of this rule.
     additional_types = [
         type_mapping[dep.label]
         for dep in ctx.rule.attr.deps
         if dep.label in type_mapping
     ]
 
-    for import_ in _extract_imports(target):
+    # FIXME: Filter this against the source files of this rule. Since we're
+    # propagating MyPy caches between aspect generated action instances, we
+    # never care about and should never have transitive files present. Only
+    # direct sources should be present, and even then only `.py` and `.pyi`
+    # files need to be present on the final MYPYPATH.
+    for import_ in _extract_imports(ctx.bin_dir.path, target):
         imports_dirs[import_] = 1
 
     pyi_files = []
     pyi_dirs = {}
+
     for dep in (ctx.rule.attr.deps + additional_types):
+        # FIXME: .pyi files from deoendencies shouldn't need to be collected for
+        # use as imports since their type stubs are already registered in a
+        # typecheck output cache, and those caches are inputs. Should be able to
+        # remove this without effect.
         if RulesPythonPyInfo in dep and hasattr(dep[RulesPythonPyInfo], "direct_pyi_files"):
             pyi_files.extend(dep[RulesPythonPyInfo].direct_pyi_files.to_list())
-            pyi_dirs |= {"%s/%s" % (ctx.bin_dir.path, imp): None for imp in _extract_imports(dep) if imp != "site-packages" and imp != "_main"}
+            # pyi_dirs |= {"%s/%s" % (ctx.bin_dir.path, imp): None for imp in _extract_imports(dep) if imp != "site-packages" and imp != "_main"}
+
         depsets.append(dep.default_runfiles.files)
+
+        # FIXME: This is a hack to achieve an `imports=[]` like effect on the
+        # PyTypeLibraryInfo, which lacks such a feature. Should just move it
+        # there and make this part of normal info collection, if the type
+        # library relocation dance is required at all which it shouldn't be.
```

Comment:
the history of `py_type_library` is to deal with typings (e.g. `foo-types` for pypi package `foo`) structure -- the contents of `foo-types` generally has a root folder `foo-types`, but for mypy we need to actually shove the `.pyi` files contained within `foo-types` in `foo`. `py_type_library` does exactly that coercion, and `PyTypeLibraryInfo` then is _exactly_ the needed path.

Reply:
Can you provide an example of a types package which behaves this way? Per https://peps.python.org/pep-0561/#stub-only-packages and the current documentation this behavior seems at odds with the documented expectations of how stubs should be packaged eve by `*-types` or `*-stubs` packages.